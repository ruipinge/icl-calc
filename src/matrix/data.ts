import { DATA_POINTS, DataPoint } from '../db';

const ATA_DEVIATION: number = 0.1; // millimetres
const CLR_DEVIATION: number = 100 * 0.001; // millimetres

export const NUM_DATA_POINTS = DATA_POINTS.length;

export interface VaultRange {
  min?: number;
  max?: number;
}

type LensSizeId = 'small' | 'medium' | 'large';
interface LensSize {
  id: LensSizeId;
  label: string;
  value: number;
}
export const LENS_SIZES: LensSize[] = [
  {
    label: '12.6 mm',
    id: 'small',
    value: 12.6
  },
  {
    label: '13.2 mm',
    id: 'medium',
    value: 13.2
  },
  {
    label: '13.7 mm',
    id: 'large',
    value: 13.7
  }
];

type MyopiaId = 'low' | 'moderate' | 'high';
interface MyopiaLevel {
  id: MyopiaId;
  label: string;
  title: string;
  min: number;
  max: number;
}
export const MYOPIA_LEVELS: MyopiaLevel[] = [
  {
    id: 'low',
    label: 'Low',
    title: 'Spherical Equivalent (SE) less than -6 dpt',
    min: -6,
    max: Infinity
  },
  {
    id: 'moderate',
    label: 'Moderate',
    title: 'Spherical Equivalent (SE) between -6 dpt and -12 dpt',
    min: -12,
    max: -6
  },
  {
    id: 'high',
    label: 'High',
    title: 'Spherical Equivalent (SE) greater than -12 dpt',
    min: -Infinity,
    max: -12
  }
];

export interface MatrixFilter {
  readonly ata: number;
  readonly clr: number;
  readonly dataPoints?: DataPoint[];
}

interface CacheEntry {
  lens: LensSize;
  myopia: MyopiaLevel;
  dataPoints?: DataPoint[];
}

export const getDataPoints = ({
  lens,
  myopia,
  dataPoints = DATA_POINTS
}: CacheEntry): DataPoint[] =>
  dataPoints.filter(
    (point) =>
      lens.value === point.iclSize &&
      myopia.min <= point.iclSe &&
      point.iclSe < myopia.max
  );

const filter = ({
  ata,
  clr,
  point
}: {
  ata: number;
  clr: number;
  point: DataPoint;
}) =>
  ata >= point.ata - ATA_DEVIATION &&
  ata <= point.ata + ATA_DEVIATION &&
  clr >= point.clr - CLR_DEVIATION && // clr in db is in millimetres
  clr <= point.clr + CLR_DEVIATION; // clr in db is in millimetres

export const filterFlatRows = ({
  ata,
  clr,
  dataPoints = DATA_POINTS
}: MatrixFilter): DataPoint[][] =>
  LENS_SIZES.reduce(
    (memoLens, lens) => [
      ...memoLens,
      ...MYOPIA_LEVELS.reduce(
        (memoMyopia, level) => [
          ...memoMyopia,
          ...[
            getDataPoints({
              lens: lens,
              myopia: level,
              dataPoints: dataPoints
            }).filter((point: DataPoint) => filter({ ata, clr, point }))
          ]
        ],
        [] as DataPoint[][]
      )
    ],
    [] as DataPoint[][]
  );
