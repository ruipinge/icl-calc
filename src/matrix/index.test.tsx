import { Matrix, getNumEyes } from '.';

import TestRenderer from 'react-test-renderer';
import { FILTER as filter } from './data.test';

it('renders without crashing', () => {
  const tree = TestRenderer.create(<Matrix ata={11.7} clr={0} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('calculates number of eyes', () => {
  const eyes = getNumEyes(filter);
  expect(eyes).toEqual([3, 3, 3, 3, 3, 3, 3, 3, 3]);
});
