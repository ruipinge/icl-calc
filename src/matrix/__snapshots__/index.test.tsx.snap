// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
Array [
  <table
    className="table table-bordered table-hover"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Lens Size
        </th>
        <th
          className="text-right"
          colSpan={3}
          scope="col"
        >
          12.6 mm
        </th>
        <th
          className="text-right"
          colSpan={3}
          scope="col"
        >
          13.2 mm
        </th>
        <th
          className="text-right"
          colSpan={3}
          scope="col"
        >
          13.7 mm
        </th>
      </tr>
      <tr>
        <th
          scope="col"
        >
          Myopia
        </th>
        <th
          className="text-right"
          scope="col"
          title="Spherical Equivalent (SE) less than -6 dpt"
        >
          Low
        </th>
        <th
          className="text-right"
          scope="col"
          title="Spherical Equivalent (SE) between -6 dpt and -12 dpt"
        >
          Moderate
        </th>
        <th
          className="text-right"
          scope="col"
          title="Spherical Equivalent (SE) greater than -12 dpt"
        >
          High
        </th>
        <th
          className="text-right"
          scope="col"
          title="Spherical Equivalent (SE) less than -6 dpt"
        >
          Low
        </th>
        <th
          className="text-right"
          scope="col"
          title="Spherical Equivalent (SE) between -6 dpt and -12 dpt"
        >
          Moderate
        </th>
        <th
          className="text-right"
          scope="col"
          title="Spherical Equivalent (SE) greater than -12 dpt"
        >
          High
        </th>
        <th
          className="text-right"
          scope="col"
          title="Spherical Equivalent (SE) less than -6 dpt"
        >
          Low
        </th>
        <th
          className="text-right"
          scope="col"
          title="Spherical Equivalent (SE) between -6 dpt and -12 dpt"
        >
          Moderate
        </th>
        <th
          className="text-right"
          scope="col"
          title="Spherical Equivalent (SE) greater than -12 dpt"
        >
          High
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
          title="Number of Eyes matching the column criteria"
        >
          Number of Eyes
        </th>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          5
        </td>
        <td
          className="text-right"
        >
          6
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          8
        </td>
        <td
          className="text-right"
        >
          6
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
      </tr>
      <tr
        className="table-bordered"
      >
        <td
          colSpan={10}
        />
      </tr>
      <tr>
        <th
          scope="row"
          title="Average Vault size in micrometres"
        >
          Average Vault (μm)
        </th>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          618
        </td>
        <td
          className="text-right"
        >
          585
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          915
        </td>
        <td
          className="text-right"
        >
          680
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
      </tr>
      <tr>
        <th
          scope="row"
          title="Minimum Vault size in micrometres"
        >
          Minimum Vault (μm)
        </th>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          350
        </td>
        <td
          className="text-right"
        >
          300
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          570
        </td>
        <td
          className="text-right"
        >
          430
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
      </tr>
      <tr>
        <th
          scope="row"
          title="Maximum Vault size in micrometres"
        >
          Maximum Vault (μm)
        </th>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          1100
        </td>
        <td
          className="text-right"
        >
          1020
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          1290
        </td>
        <td
          className="text-right"
        >
          880
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
      </tr>
      <tr
        className="table-bordered"
      >
        <td
          colSpan={10}
        />
      </tr>
      <tr>
        <th
          scope="row"
          title="Percentage of Eyes with Vault size less than 250 micrometres"
        >
          % Vault &lt; 250 (μm)
        </th>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
      </tr>
      <tr>
        <th
          scope="row"
          title="Percentage of Eyes with Vault size between 250 and 500 micrometres"
        >
          % 250 &lt; Vault &lt; 500 (μm)
        </th>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0.4
        </td>
        <td
          className="text-right"
        >
          0.333
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0.167
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
      </tr>
      <tr>
        <th
          scope="row"
          title="Percentage of Eyes with Vault size between 500 and 750 micrometres"
        >
          % 500 &lt; Vault &lt; 750 (μm)
        </th>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0.4
        </td>
        <td
          className="text-right"
        >
          0.333
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0.375
        </td>
        <td
          className="text-right"
        >
          0.333
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
      </tr>
      <tr>
        <th
          scope="row"
          title="Percentage of Eyes with Vault size between 750 and 1000 micrometres"
        >
          % 750 &lt; Vault &lt; 1000 (μm)
        </th>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0.167
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0.25
        </td>
        <td
          className="text-right"
        >
          0.5
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
      </tr>
      <tr>
        <th
          scope="row"
          title="Percentage of Eyes with Vault size greater or equal than 1000 micrometres"
        >
          % 1000 &lt; Vault (μm)
        </th>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0.2
        </td>
        <td
          className="text-right"
        >
          0.167
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0.375
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
        <td
          className="text-right"
        >
          0
        </td>
      </tr>
    </tbody>
  </table>,
  <ul
    className="list-inline"
  >
    <li
      className="list-inline-item"
    >
      <strong>
        Angle to Angle (AtA): 
      </strong>
      11.7
       mm.
    </li>
    <li
      className="list-inline-item"
    >
      <strong>
        Crystaline Lens Rise (CLR): 
      </strong>
      0
       mm.
    </li>
    <li
      className="list-inline-item"
    >
      <strong>
        Number of matching Eyes: 
      </strong>
      25
      /
      542
      .
    </li>
  </ul>,
]
`;
